### PR TITLE
Adds Etsy to OAuth providers

### PR DESCRIPTION
**Pull request by wess**
## What does this PR do?

Adds Etsy as an OAuth Provider

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work.)

Add application and enable in [Etsy](https://developer.etsy.com)

## Related PRs and Issues

(If this PR is related to any other PR or resolves any issue or related to any issue link all related PR and issues here.)

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

Yes

### CODE REVIEW

**Comment by TorstenDittmann** [comment]
@wess linter is failing 👍🏻

File: .phpunit.result.cache [hunk 1]
```diff
@@ -0,0 +1 @@
...ustomServerTest::testListTags":0,"Tests\\E2E\\Services\\Functions\\FunctionsCustomServerTest::testGetTag":0,"Tests\\E2E\\Services\\Functions\\FunctionsCustomServerTest::testCreateExecution":0,"Tests\\E2E\\Services\\Functions\\FunctionsCustomServerTest::testListExecutions":0,"Tests\\E2E\\Services\\Functions\\FunctionsCustomServerTest::testGetExecution":0,"Tests\\E2E\\Services\\Functions\\FunctionsCustomServerTest::testDeleteTag":0,"Tests\\E2E\\Services\\Functions\\FunctionsCustomServerTest::testDelete":0,"Tests\\E2E\\Services\\Functions\\FunctionsCustomServerTest::testTimeout":0.001,"Tests\\E2E\\Services\\Functions\\FunctionsCustomServerTest::testCreateCustomExecution":0,"Tests\\E2E\\Services\\Functions\\FunctionsCustomServerTest::testGetRuntimes":0.001,"Tests\\E2E\\Services\\Functions\\FunctionsCustomClientTest::testCreate":0.001,"Tests\\E2E\\Services\\Functions\\FunctionsCustomClientTest::testCreateExecution":0,"Tests\\E2E\\Services\\Functions\\FunctionsCustomClientTest::testCreateCustomExecution":0,"Tests\\E2E\\Services\\Functions\\FunctionsCustomClientTest::testCreateExecutionUnauthorized":0.001,"Tests\\E2E\\Services\\Functions\\FunctionsCustomClientTest::testListExecutions":0}}
```
Comment (TorstenDittmann) [inline .phpunit.result.cache:1]: Can we remove that file and add it to `.gitignore`?

File: src/Appwrite/Auth/OAuth2/etsy.php [hunk 2]
```diff
@@ -0,0 +1,196 @@
+<?php
+
+namespace Appwrite\Auth\OAuth2;
+
+use Appwrite\Auth\OAuth2;
+
+class Notion extends OAuth2
```
Comment (TorstenDittmann) [inline src/Appwrite/Auth/OAuth2/etsy.php:7]:
```suggestion
class Etsy extends OAuth2
```

File: src/Appwrite/Auth/OAuth2/etsy.php [hunk 3]
```diff
@@ -0,0 +1,212 @@
+<?php
+
```
Comment (eldadfux) [inline src/Appwrite/Auth/OAuth2/etsy.php:2]: @wess we need to remove this file.

File: src/Appwrite/Auth/OAuth2/Etsy.php [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+namespace Appwrite\Auth\OAuth2;
+
+use Appwrite\Auth\OAuth2;
+
+class Etsy extends OAuth2
+{
+    /**
+     * @var string
+     */
+    private $endpoint = 'https://api.etsy.com/v3/public';
```
Comment (TorstenDittmann) [inline src/Appwrite/Auth/OAuth2/Etsy.php:12]: We have to add typings here and everywhere else 👍🏻 
```suggestion
    private string $endpoint = 'https://api.etsy.com/v3/public';
```

File: src/Appwrite/Auth/OAuth2/Etsy.php [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        "listings_r",
+        "listings_w",
+        "profile_r",
+        "profile_w",
+        "recommend_r",
+        "recommend_w",
+        "shops_r",
+        "shops_w",
+        "transactions_r",
+        "transactions_w",
```
Comment (TorstenDittmann) [inline src/Appwrite/Auth/OAuth2/Etsy.php:52]: are all of those necessary to work?

Here should only be the bare minimum for it to work 👍🏻

File: src/Appwrite/Auth/OAuth2/Etsy.php [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+     */
+    public function getUserEmail(string $accessToken): string
+    {
+        return $this->getUser($accessToken)['primary_email'];
+    }
+
+    /**
+     * Check if the OAuth email is verified
+     *
+     * If present, the email is verified. This was verfied through a manual Stripe sign up process
```
Comment (TorstenDittmann) [inline src/Appwrite/Auth/OAuth2/Etsy.php:169]: This comment is wrong.

Comment (TorstenDittmann) [reply src/Appwrite/Auth/OAuth2/Etsy.php:169]: Can we double check that this is the case here as well?

**Review by christyjacob4 — APPROVED** [review]

File: src/Appwrite/Auth/OAuth2/Etsy.php [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+    {
+        return $this->getUser($accessToken)['login_name'];
+    }
+
+    /**
+     * @param string $accessToken
+     *
+     * @return array
+     */
+    protected function getUser(string $accessToken)
```
Comment (christyjacob4) [inline src/Appwrite/Auth/OAuth2/Etsy.php:179]:
```suggestion
    protected function getUser(string $accessToken): array
```

File: src/Appwrite/Auth/OAuth2/Etsy.php [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+{
+    /**
+     * @var string
+     */
+    private string $endpoint = 'https://api.etsy.com/v3/public';
+
+    /**
+     * @var string
+     */
+    private $version = '2022-07-14';
```
Comment (christyjacob4) [inline src/Appwrite/Auth/OAuth2/Etsy.php:17]:
```suggestion
    private string $version = '2022-07-14';
```
As @TorstenDittmann  suggested, can you please add typings to all the variables ?

File: src/Appwrite/Auth/OAuth2/Etsy.php [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+
+    /**
+     * @var array
+     */
+    protected array $scopes = [
+        "email_r",
+        "profile_r",
+    ];
+
+    private $pkce = '';
```
Comment (christyjacob4) [inline src/Appwrite/Auth/OAuth2/Etsy.php:37]:
```suggestion
    /**
     * @var string
     */
    private string $pkce = '';
```

File: src/Appwrite/Auth/OAuth2/Etsy.php [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+
+    private function getPKCE()
```
Comment (christyjacob4) [inline src/Appwrite/Auth/OAuth2/Etsy.php:39]:
```suggestion
    private function getPKCE(): string
```

**Review by christyjacob4 — CHANGES_REQUESTED** [review]
Please address the comments

File: app/config/providers.php [hunk 11]
```diff
@@ -101,6 +101,16 @@
         'beta' => false,
         'mock' => false,
     ],
+    'Etsy' => [
```
Comment (christyjacob4) [inline app/config/providers.php:104]:
```suggestion
    'etsy' => [
```

The Keys for the providers are expected to be lower cased 👀

File: src/Appwrite/Auth/OAuth2/Etsy.php [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+
+    /**
+     * @var string
+     */
+    private string $version = '2022-07-14';
+
+    /**
+     * @var array
+     */
+    protected $user = [];
```
Comment (christyjacob4) [inline src/Appwrite/Auth/OAuth2/Etsy.php:22]:
```suggestion
    protected array $user = [];
```

File: src/Appwrite/Auth/OAuth2/Etsy.php [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+
+    /**
+     * @var array
+     */
+    protected $tokens = [];
```
Comment (christyjacob4) [inline src/Appwrite/Auth/OAuth2/Etsy.php:27]:
```suggestion
    protected array $tokens = [];
```

File: src/Appwrite/Auth/OAuth2/Etsy.php [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+        "email_r",
+        "profile_r",
+    ];
+
+    /**
+     * @var string
+     */
+    private string $pkce = '';
+
+    private function getPKCE(): string
```
Comment (christyjacob4) [inline src/Appwrite/Auth/OAuth2/Etsy.php:42]:
```suggestion
    /**
     * @return string
     */
    private function getPKCE(): string
```

**Review by christyjacob4 — CHANGES_REQUESTED** [review]
Looks good. Just a few more comments

**Pull request merged**